src/test/result_tests.cpp

```diff
@@ -45,42 +74,122 @@ util::Result<bilingual_str> StrFn(bilingual_str s, bool success)
     return util::Error{strprintf(Untranslated("str %s error."), s.original)};
 }
 
-util::Result<NoCopy> NoCopyFn(int i, bool success)
+util::Result<NoCopy, NoCopy> NoCopyFn(int i, bool success)
 {
     if (success) return {i};
-    return util::Error{Untranslated(strprintf("nocopy %i error.", i))};
+    return {util::Error{Untranslated(strprintf("nocopy %i error.", i))}, i};
 }
 
-template <typename T>
-void ExpectResult(const util::Result<T>& result, bool success, const bilingual_str& str)
+util::Result<NoCopyNoMove, NoCopyNoMove> NoCopyNoMoveFn(int i, bool success)
+{
+    if (success) return {i};
+    return {util::Error{Untranslated(strprintf("nocopynomove %i error.", i))}, i};
+}
+
+enum FnError { ERR1, ERR2 };
+
+util::Result<int, FnError> IntFailFn(int i, bool success)
+{
+    if (success) return i;
+    return {util::Error{Untranslated(strprintf("int %i error.", i))}, i % 2 ? ERR1 : ERR2};
+}
+
+util::Result<NoCopyNoMove, FnError> EnumFailFn(FnError ret)
+{
+    return {util::Error{Untranslated("enum fail.")}, ret};
+}
+
+util::Result<void> WarnFn()
+{
+    return {util::Warning{Untranslated("warn.")}};
+}
+
+util::Result<int> MultiWarnFn(int ret)
+{
+    util::Result<void> result;
+    for (int i = 0; i < ret; ++i) {
+        result.AddWarning(strprintf(Untranslated("warn %i."), i));
+    }
+    return {std::move(result), ret};
+}
+
+util::Result<void, int> ChainedFailFn(FnError arg, int ret)
+{
+    return {util::Error{Untranslated("chained fail.")}, EnumFailFn(arg), WarnFn(), ret};
+}
+
+util::Result<int, FnError> AccumulateFn(bool success)
+{
+    util::Result<int, FnError> result;
+    util::Result<int> x = MultiWarnFn(1) >> result;
+    BOOST_REQUIRE(x);
+    util::Result<int> y = MultiWarnFn(2) >> result;
+    BOOST_REQUIRE(y);
+    result.Set(IntFailFn(*x + *y, success));
+    return result;
+}
+
+util::Result<int, int> TruthyFalsyFn(int i, bool success)
+{
+    if (success) return i;
+    return {util::Error{Untranslated(strprintf("failure value %i.", i))}, i};
+}
+
```

Comment:
Can you add an example for when the Result error type in a chain of function calls remains the same, but the value type changes? Is there a better way to do this conversion than manually constructing the result again (like below)?

```suggestion
util::Result<std::string, FnError> CastFailFn() {
    auto res = IntFailFn(1, false);
    return {util::Error{ErrorString(res)}, res.GetFailure()};
}
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r1234264247

> Can you add an example for when the Result error type in a chain of function calls remains the same, but the value type changes? Is there a better way to do this conversion than manually constructing the result again (like below)?

Yes, there is definitely a better way to pass along errors and warnings from one result to another without calling `ErrorString` and flattening them. The `util::Result` constructor will move errors and warnings directly from an existing result object if you just pass the other result object as an argument with `std::move`. If you look at the `MultiWarnFn` there was an example of this feature in the success case. But I added a new `StrFailFn` function now based on your code that demonstrates it in both the success and failure cases.

I am also thinking of adding a `util::Messages{Result&&}` helper similar to existing `util::Error{std::string}` and `util::Warning{std::string}` helpers to make it more obvious how you can construct a Result value with errors and warning from different sources.

Revisiting this makes me realize I should probably add a tutorial-style document that describes how to use the Result class with standalone functions that return util::Result, chained functions that return util::Result values of the same type, and chained functions that return util::Result values of different types.